crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1436,76 +1444,109 @@ impl<'db> Type<'db> {
+                                    }
+                                    // The type has a `__bool__` method, but it doesn't return a boolean.
+                                    _ => {
+                                        return Err(BoolError::BindingError {
+                                            return_type: outcome.return_type(db),
+                                            not_boolable_type: *instance_ty,
+                                        });
+                                    }
+                                },
+
+                                Err(CallError::BindingError { binding }) => {
+                                    return Err(BoolError::BindingError {
+                                        return_type: binding.return_type(),
+                                        not_boolable_type: *instance_ty,
+                                    });
+                                }
+                                Err(CallError::NotCallable { .. }) => {
+                                    return Err(BoolError::NotCallable {
+                                        not_boolable_type: *instance_ty,
+                                    })
+                                }
+                                Err(
+                                    CallError::PossiblyUnboundDunderCall { .. }
+                                    | CallError::Union { .. },
+                                ) => {
+                                    return Err(BoolError::Other {
+                                        not_boolable_type: *instance_ty,
+                                    })
+                                }
+                            },
+                            // Don't trust possibly unbound `__bool__` method.
+                            Symbol::Type(_, Boundness::PossiblyUnbound) => Truthiness::Ambiguous,
+                            Symbol::Unbound => Truthiness::Ambiguous,
+                        }
                     }
                 }
-            }
-            Type::KnownInstance(known_instance) => known_instance.bool(),
-            Type::Union(union) => {
-                let union_elements = union.elements(db);
-                let first_element_truthiness = union_elements[0].bool(db);
-                if first_element_truthiness.is_ambiguous() {
-                    return Truthiness::Ambiguous;
+                Type::KnownInstance(known_instance) => known_instance.bool(),
+                Type::Union(union) => {
+                    let mut union_elements = union.elements(db).iter();
+                    let first_element = union_elements.next().unwrap();
+                    let mut truthiness = first_element.bool(db)?;
+
+                    for element in union_elements {
+                        if element.bool(db)? != truthiness {
+                            truthiness = Truthiness::Ambiguous;
+                        }
+                    }
```

Comment:
I suspect that the regression comes from the fact that we no longer short-circuit if any union element inferred `Ambigious`.